### PR TITLE
add get_top_markets

### PR DESCRIPTION
**Pull request by oxarbitrage**
replaces https://github.com/bitshares/bitshares-core/pull/560

Edited by @abitmore: this is new PR for https://github.com/bitshares/bitshares-core/issues/512.

### CODE REVIEW

File: libraries/app/database_api.cpp [hunk 1]
```diff
@@ -1292,6 +1293,40 @@ order_book database_api_impl::get_order_book( const string& base, const string&
...
+vector<market_volume> database_api_impl::get_top_markets(uint limit)const
+{
+   FC_ASSERT( limit <= 100 );
+
+   const auto& volume_idx = _db.get_index_type<graphene::market_history::market_ticker_index>().indices().get<by_volume>();
+   auto itr = volume_idx.rbegin();
+   vector<market_volume> result;
+
+   fc::uint128 base_volume;
+   fc::uint128 quote_volume;
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: Unused variables?

File: libraries/app/database_api.cpp [hunk 2]
```diff
@@ -1292,6 +1293,40 @@ order_book database_api_impl::get_order_book( const string& base, const string&
...
+
+   uint c = 0;
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: Unused variable?

File: libraries/app/database_api.cpp [hunk 3]
```diff
@@ -115,6 +115,7 @@ class database_api_impl : public std::enable_shared_from_this<database_api_impl>
       market_ticker                      get_ticker( const string& base, const string& quote, bool skip_order_book = false )const;
       market_volume                      get_24_volume( const string& base, const string& quote )const;
       order_book                         get_order_book( const string& base, const string& quote, unsigned limit = 50 )const;
+      vector<market_volume>              get_top_markets(uint limit)const;
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: Code style: add spaces.

By the way, I don't like `uint`, perhaps change to `uint32_t`?

**Comment by oxarbitrage** [comment]
cleaned that, thanks.

File: libraries/app/database_api.cpp [hunk 4]
```diff
@@ -1292,6 +1293,36 @@ order_book database_api_impl::get_order_book( const string& base, const string&
...
+   while( itr != volume_idx.rend() && result.size() < limit)
+   {
+      market_volume mv;
+      mv.time = now;
+      const auto assets = get_assets( { itr->base, itr->quote } );
+      mv.base = assets[0]->symbol;
+      mv.quote = assets[1]->symbol;
+      mv.base_volume = uint128_amount_to_string( itr->base_volume, assets[0]->precision );
+      mv.quote_volume = uint128_amount_to_string( itr->quote_volume, assets[1]->precision );
+      result.push_back(mv);
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: I think `result.emplace_back( std::move(mv) )` would have better performance.

File: libraries/app/database_api.cpp [hunk 5]
```diff
@@ -1292,6 +1293,36 @@ order_book database_api_impl::get_order_book( const string& base, const string&
...
+   return my->get_top_markets(limit);
+}
+
+vector<market_volume> database_api_impl::get_top_markets(uint32_t limit)const
+{
+   FC_ASSERT( limit <= 100 );
+
+   const auto& volume_idx = _db.get_index_type<graphene::market_history::market_ticker_index>().indices().get<by_volume>();
+   auto itr = volume_idx.rbegin();
+   vector<market_volume> result;
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: Adding a `result.reserve(limit);` here would improve performance.

**Comment by oxarbitrage** [comment]
thank you, added the performance changes at https://github.com/bitshares/bitshares-core/pull/737/commits/eaefc599c908bd54e88124833bb7253c5e47b0ed

**Comment by oxarbitrage** [comment]
agree, added and merged.

**Pull request merged**